python-package/lightgbm/basic.py

```diff
@@ -1853,6 +1853,119 @@ def free_network(self):
+                    'missing_direction': None,
+                    'missing_type': None,
+                    'internal_value': None,
+                    'leaf_value': None,
+                    'weight': None,
+                    'count': None}
+
+            # Update values to reflect node type (leaf or split)
+            if _is_split_node(tree):
+                node['left_child'] = _get_node_index(tree['left_child'], tree_index)
+                node['right_child'] = _get_node_index(tree['right_child'], tree_index)
+                node['split_gain'] = tree['split_gain']
+                node['threshold'] = tree['threshold']
+                node['decision_type'] = tree['decision_type']
+                node['missing_direction'] = 'left' if tree['default_left'] else 'right'
+                node['missing_type'] = tree['missing_type']
+                node['internal_value'] = tree['internal_value']
+                node['weight'] = tree['internal_weight']
+                node['count'] = tree['internal_count']
+            else:
+                node['leaf_value'] = tree['leaf_value']
+                node['weight'] = tree['leaf_weight']
+                node['count'] = tree['leaf_count']
+
+            return node
+
+        def tree_dict_to_node_list(tree, node_depth=1, tree_index=None,
+                                   feature_names=None, parent_node=None):
+
+            node = create_node_record(tree,
+                                      node_depth=node_depth,
+                                      tree_index=tree_index,
+                                      feature_names=feature_names,
+                                      parent_node=parent_node)
+
+            res = [node]
+
+            if _is_split_node(tree):
+                # traverse the next level of the tree
+                children = ['left_child', 'right_child']
+                for child in children:
+                    subtree_list = tree_dict_to_node_list(
+                        tree[child],
+                        node_depth=node_depth + 1,
+                        tree_index=tree_index,
+                        feature_names=feature_names,
+                        parent_node=node['node_index'])
+                    # In tree format, "subtree_list" is a list of node records (dicts),
+                    # and we add node to the list.
+                    res.extend(subtree_list)
+            return res
+
+        model_dict = self.dump_model()
+        feature_names = model_dict['feature_names']
+        model_list = []
+        for tree in model_dict['tree_info']:
+            model_list.extend(tree_dict_to_node_list(tree['tree_structure'],
+                                                     tree_index=tree['tree_index'],
+                                                     feature_names=feature_names))
+
+        if PANDAS_INSTALLED:
+            return DataFrame(model_list, columns=model_list[0].keys())
+        else:
+            return model_list
```

Comment:
Just a question: what are the cases when this kind of output is useful? To be a source for any other dataframe-like object or what for? In other words, in what terms it's better than raw output from `dump_model()`? I'm asking because it contradicts with the method name a little bit.

Reply:
This is very convenient way to look at things like maximum and minimum split gains which is useful for selecting hyperparamters or ranges of hyperparameters such as `min_gain_to_split`. You would have to write a lot of custom code to that from `dump_model()` output, but with pandas it's very easy. Also, it's a convenient way to see how much trees are being pruned, how imbalanced they are, etc. Finally, it's useful for figuring out which variables tend to have "interaction" effects which can tell you something about your data. All of those things are much more difficult when the data is in nested key:value structure like from `dump.model()`.

Reply:
Thank you very much for providing a rationale of this PR, but I was referring to a case when there is no pandas installed: `return model_list`. Shouldn't we simply raise an error when pandas was not detected?

> ... but with pandas it's very easy.

Reply:
Ah, I see. I agree, I don't think a list format would be useful at all. In the latest commit I'm not raising a LightGBM exception if pandas is not installed.